File: app/handler/video.go
package handler

import (
	"cczjVideo/app/db"
	"cczjVideo/app/model"
	"cczjVideo/app/util"
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
)

// parseInt64 同时接受数字和字符串形式的数字
func parseInt64(raw json.RawMessage, def int64) (int64, error) {
	if len(raw) == 0 {
		return def, nil
	}
	// 字符串形式："0"
	if raw[0] == '"' {
		var s string
		if err := json.Unmarshal(raw, &s); err != nil {
			return def, err
		}
		if s == "" {
			return def, nil
		}
		v, err := strconv.ParseInt(s, 10, 64)
		if err != nil {
			return def, fmt.Errorf("parse %q: %w", s, err)
		}
		return v, nil
	}
	var v int64
	if err := json.Unmarshal(raw, &v); err != nil {
		return def, err
	}
	return v, nil
}

func parseInt(raw json.RawMessage, def int) (int, error) {
	v, err := parseInt64(raw, int64(def))
	if err != nil {
		return def, err
	}
	return int(v), nil
}

type VideoListReq struct {
	SourceKey string `json:"source_key"`
	// type_id 同时支持字符串与数字，保存为字符串
	TypeId   string `json:"type_id"`
	Year     string `json:"year"`     // 年份筛选（"all" 或具体年份）
	Area     string `json:"area"`     // 地区筛选（"all" 或具体地区）
	Keyword  string `json:"keyword"`  // 关键词：标题/演员/导演/备注/年份/地区/类型 模糊匹配
	Sort     string `json:"sort"`     // "" 默认; "rating" 按评分; "hot" 按热度
	Page     int    `json:"page"`
	PageSize int    `json:"page_size"`
}

func (r *VideoListReq) UnmarshalJSON(data []byte) error {
	var raw struct {
		SourceKey string          `json:"source_key"`
		TypeId    json.RawMessage `json:"type_id"`
		Year      string          `json:"year"`
		Area      string          `json:"area"`
		Keyword   string          `json:"keyword"`
		Sort      string          `json:"sort"`
		Page      json.RawMessage `json:"page"`
		PageSize  json.RawMessage `json:"page_size"`
	}
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	r.SourceKey = raw.SourceKey
	r.TypeId = normalizeStringId(raw.TypeId)
	r.Year = strings.TrimSpace(raw.Year)
	r.Area = strings.TrimSpace(raw.Area)
	r.Keyword = strings.TrimSpace(raw.Keyword)
	r.Sort = strings.TrimSpace(raw.Sort)
	if v, err := parseInt(raw.Page, 1); err == nil {
		r.Page = v
	}
	if v, err := parseInt(raw.PageSize, 20); err == nil {
		r.PageSize = v
	}
	return nil
}

// normalizeStringId 将 json.RawMessage 规范化为字符串（支持字符串/数字/空）
func normalizeStringId(raw json.RawMessage) string {
	if len(raw) == 0 {
		return ""
	}
	// 带引号的字符串
	if raw[0] == '"' {
		var s string
		if err := json.Unmarshal(raw, &s); err == nil {
			return s
		}
		return ""
	}
	// 数字：直接转为字符串
	var n int64
	if err := json.Unmarshal(raw, &n); err == nil {
		return strconv.FormatInt(n, 10)
	}
	var f float64
	if err := json.Unmarshal(raw, &f); err == nil {
		return strconv.FormatInt(int64(f), 10)
	}
	return ""
}

type VideoListResp struct {
	Videos []*model.Video `json:"videos"`
	Total  int            `json:"total"`
}

func GetVideoList(req VideoListReq) (*VideoListResp, error) {
	if req.Page <= 0 {
		req.Page = 1
	}
	if req.PageSize <= 0 {
		req.PageSize = 20
	}

	filter := db.FilterParams{
		TypeId:   req.TypeId,
		Year:     req.Year,
		Area:     req.Area,
		Keyword:  req.Keyword,
		Sort:     req.Sort,
		Page:     req.Page,
		PageSize: req.PageSize,
	}
	videos, total, err := db.GetVideos(req.SourceKey, filter)
	if err != nil {
		return nil, fmt.Errorf("get videos: %w", err)
	}

	for _, v := range videos {
		v.VodActor = util.DecompressIfNeeded(v.VodActor)
		v.VodDirector = util.DecompressIfNeeded(v.VodDirector)
		v.VodContent = util.DecompressIfNeeded(v.VodContent)
	}
	// 用全局豆瓣数据批量补充缺失字段（一次 JOIN 替代 N×2 次查询）
	db.EnrichVideosWithDouban(videos)

	return &VideoListResp{Videos: videos, Total: total}, nil
}

// YearAreaResp 前端用于前端：返回所有可选年份 / 地区列表（用于筛选下拉框的选项）
type YearsResp struct {
	Years []string `json:"years"`
	Areas []string `json:"areas"`
}

func GetYearsAndAreas(sourceKey string) (*YearsResp, error) {
	years, _ := db.GetDistinctYears(sourceKey)
	areas, _ := db.GetDistinctAreas(sourceKey)
	return &YearsResp{Years: years, Areas: areas}, nil
}

// GetRecommend 返回 N 条推荐视频（排除指定 id 集合中的视频）
func GetRecommend(sourceKey string, limit int, excludeIds []string) ([]*model.Video, error) {
	videos, err := db.GetRandomRecommend(sourceKey, limit, excludeIds)
	if err != nil {
		return nil, err
	}
	for _, v := range videos {
		v.VodActor = util.DecompressIfNeeded(v.VodActor)
		v.VodDirector = util.DecompressIfNeeded(v.VodDirector)
	}
	// 用全局豆瓣数据批量补充缺失字段（一次 JOIN 替代 N×2 次查询）
	db.EnrichVideosWithDouban(videos)
	return videos, nil
}

// HistoryItemWithVideo 前端可用的"继续观看"条目：含视频名+封面，便于卡片展示
type HistoryItemWithVideo struct {
	SourceKey string  `json:"source_key"`
	VodId     string  `json:"vod_id"`
	EpNum     int     `json:"ep_num"`
	Position  float64 `json:"position"`
	UpdatedAt string  `json:"updated_at"`
	VodName   string  `json:"vod_name"`
	VodPic    string  `json:"vod_pic"`
	VodRemarks string `json:"vod_remarks"`
}

// HydrateHistory hydrates raw history entries with additional video info from global_video
func HydrateHistory(sourceKey string, raws []db.HistEntry) []*HistoryItemWithVideo {
	if len(raws) == 0 {
		return []*HistoryItemWithVideo{}
	}
	out := make([]*HistoryItemWithVideo, 0, len(raws))
	for _, r := range raws {
		item := &HistoryItemWithVideo{
			SourceKey: r.SourceKey,
			VodId:     r.VodId,
			EpNum:     r.EpNum,
			Position:  r.Position,
			UpdatedAt: r.UpdatedAt,
			VodName:   r.VodName,
			VodPic:    r.VodPic,
		}
		// Get additional info from source video table if available
		if v, err := db.GetVideoById(r.SourceKey, r.VodId); err == nil && v != nil {
			if item.VodName == "" {
				item.VodName = v.VodName
			}
			if item.VodPic == "" {
				item.VodPic = v.VodPic
			}
			item.VodRemarks = v.VodRemarks
		}
		out = append(out, item)
	}
	return out
}

type VideoDetailReq struct {
	SourceKey string `json:"source_key"`
	VodId     string `json:"vod_id"`
}

func (r *VideoDetailReq) UnmarshalJSON(data []byte) error {
	var raw struct {
		SourceKey string          `json:"source_key"`
		VodId     json.RawMessage `json:"vod_id"`
	}
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	r.SourceKey = raw.SourceKey
	r.VodId = normalizeStringId(raw.VodId)
	return nil
}

type VideoDetailResp struct {
	Video    *model.Video     `json:"video"`
	Episodes []*model.Episode `json:"episodes"`
}

func GetVideoDetail(req VideoDetailReq) (*VideoDetailResp, error) {
	video, err := db.GetVideoById(req.SourceKey, req.VodId)
	if err != nil {
		return nil, fmt.Errorf("video not found: %w", err)
	}

	video.VodContent = util.DecompressIfNeeded(video.VodContent)
	video.VodActor = util.DecompressIfNeeded(video.VodActor)
	video.VodDirector = util.DecompressIfNeeded(video.VodDirector)
	video.VodPlayUrl = util.DecompressIfNeeded(video.VodPlayUrl)
	video.VodDownUrl = util.DecompressIfNeeded(video.VodDownUrl)

	// 用全局豆瓣数据补充缺失字段
	db.EnrichVideoWithDouban(video)

	episodes, _ := db.GetEpisodes(req.SourceKey, req.VodId)

	return &VideoDetailResp{Video: video, Episodes: episodes}, nil
}

type VideoSearchReq struct {
	SourceKey string `json:"source_key"`
	Keyword   string `json:"keyword"`
	Page      int    `json:"page"`
	PageSize  int    `json:"page_size"`
}

func (r *VideoSearchReq) UnmarshalJSON(data []byte) error {
	var raw struct {
		SourceKey string          `json:"source_key"`
		Keyword   string          `json:"keyword"`
		Page      json.RawMessage `json:"page"`
		PageSize  json.RawMessage `json:"page_size"`
	}
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	r.SourceKey = raw.SourceKey
	r.Keyword = raw.Keyword
	if v, err := parseInt(raw.Page, 1); err == nil {
		r.Page = v
	}
	if v, err := parseInt(raw.PageSize, 20); err == nil {
		r.PageSize = v
	}
	return nil
}

func SearchVideos(req VideoSearchReq) (*VideoListResp, error) {
	if req.Page <= 0 {
		req.Page = 1
	}
	if req.PageSize <= 0 {
		req.PageSize = 20
	}

	videos, total, err := db.SearchVideos(req.SourceKey, req.Keyword, req.Page, req.PageSize)
	if err != nil {
		return nil, fmt.Errorf("search videos: %w", err)
	}

	for _, v := range videos {
		v.VodActor = util.DecompressIfNeeded(v.VodActor)
		v.VodDirector = util.DecompressIfNeeded(v.VodDirector)
		v.VodContent = util.DecompressIfNeeded(v.VodContent)
	}
	// 用全局豆瓣数据批量补充缺失字段（一次 JOIN 替代 N×2 次查询）
	db.EnrichVideosWithDouban(videos)
	return &VideoListResp{Videos: videos, Total: total}, nil
}

type GetTypesReq struct {
	SourceKey string `json:"source_key"`
}

func GetTypes(req GetTypesReq) ([]*model.VType, error) {
	return db.GetTypes(req.SourceKey)
}

type DeleteVideoReq struct {
	SourceKey string `json:"source_key"`
	VodId     string `json:"vod_id"`
}

func DeleteVideo(req DeleteVideoReq) error {
	if req.SourceKey == "" || req.VodId == "" {
		return fmt.Errorf("参数不完整")
	}
	return db.DeleteVideo(req.SourceKey, req.VodId)
}

File: frontend/src/stores/video.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { GetVideoList, SearchVideos, GetTypes, GetYearsAndAreas, GetRecommend } from '../../bindings/cczjVideo/app'
import * as AppMod from '../../bindings/cczjVideo/app'
import type { Video, Episode, VType, VideoDetailResponse } from '../types'
import { useErrorStore } from './error'

export interface VideoFilter {
  type_id: string | number
  year: string
  area: string
  keyword: string
  sort?: string // '' = 默认; 'rating' = 按评分; 'hot' = 按热度
}

export const useVideoStore = defineStore('video', () => {
  const videos = ref<Video[]>([])
  const currentVideo = ref<Video | null>(null)
  const episodes = ref<Episode[]>([])
  const types = ref<VType[]>([])
  const years = ref<string[]>([])
  const areas = ref<string[]>([])
  const total = ref(0)
  const page = ref(1)
  const loading = ref(false)

  // 删除通知：当某页面删除视频后，其他列表页监听此值来移除该视频
  const deletedVodId = ref<string | null>(null)
  const deletedSourceKey = ref<string | null>(null)
  function notifyDeletion(sourceKey: string, vodId: string): void {
    deletedSourceKey.value = sourceKey
    deletedVodId.value = vodId
  }
  function clearDeletionNotify(): void {
    deletedSourceKey.value = null
    deletedVodId.value = null
  }

  // 全局刷新通知：当数据被清空或重新采集后，通知所有页面刷新数据
  const refreshTrigger = ref(0)
  function notifyRefresh(): void {
    refreshTrigger.value++
  }

  const errorStore = useErrorStore()

  async function loadVideos(sourceKey: string, filter: VideoFilter, p = 1, pageSize = 50): Promise<void> {
    loading.value = true
    try {
      // 内部对象：强制类型断言避免 handler.VideoListReq 新字段（year/area/keyword/sort）造成 TS 报错
      const req = {
        source_key: sourceKey,
        type_id: filter.type_id === undefined || filter.type_id === null ? '' : String(filter.type_id),
        year: filter.year ?? '',
        area: filter.area ?? '',
        keyword: filter.keyword ?? '',
        sort: filter.sort ?? '',
        page: p,
        page_size: pageSize,
      } as any
      const resp = (await GetVideoList(req)) as any
      const list: Video[] = Array.isArray(resp?.videos) ? resp.videos : []
      const ttl: number = typeof resp?.total === 'number' ? resp.total : list.length
      if (p === 1) {
        // 先加载完新数据，再一次性替换，避免短暂空白引起布局跳动
        videos.value = list
      } else {
        videos.value.push(...list)
      }
      total.value = ttl
      page.value = p
    } catch (e: any) {
      errorStore.fromError('加载视频列表失败', e, 'videoStore.loadVideos')
    } finally {
      loading.value = false
    }
  }

  async function loadDetail(sourceKey: string, vodId: string, refresh = false): Promise<void> {
    loading.value = true
    try {
      const resp = (await (AppMod as any).GetVideoDetail({ source_key: sourceKey, vod_id: vodId, refresh })) as VideoDetailResponse
      currentVideo.value = resp?.video ?? null
      episodes.value = Array.isArray(resp?.episodes) ? resp.episodes : []
    } catch (e: any) {
      const msg = e?.message || ''
      if (msg.includes('video not found') || msg.includes('sql: no rows')) {
        currentVideo.value = null
        episodes.value = []
      } else {
        errorStore.fromError('加载视频详情失败', e, 'videoStore.loadDetail')
        currentVideo.value = null
        episodes.value = []
      }
    } finally {
      loading.value = false
    }
  }

  /** 后台刷新详情（不设置 loading，用于已有本地数据后异步更新） */
  async function refreshDetail(sourceKey: string, vodId: string): Promise<boolean> {
    try {
      const resp = (await (AppMod as any).GetVideoDetail({ source_key: sourceKey, vod_id: vodId, refresh: true })) as VideoDetailResponse
      if (resp?.video) {
        currentVideo.value = resp.video
        if (Array.isArray(resp.episodes) && resp.episodes.length > 0) {
          episodes.value = resp.episodes
        }
        return true
      }
      return false
    } catch {
      return false
    }
  }

  async function search(sourceKey: string, keyword: string, p = 1): Promise<void> {
    loading.value = true
    try {
      const resp = (await SearchVideos({
        source_key: sourceKey,
        keyword,
        page: p,
        page_size: 50,
      })) as any
      const list: Video[] = Array.isArray(resp?.videos) ? resp.videos : []
      const ttl: number = typeof resp?.total === 'number' ? resp.total : list.length
      if (p === 1) {
        videos.value = list
      } else {
        videos.value.push(...list)
      }
      total.value = ttl
      page.value = p
    } catch (e: any) {
      errorStore.fromError('搜索失败', e, 'videoStore.search')
    } finally {
      loading.value = false
    }
  }

  async function loadTypes(sourceKey: string): Promise<void> {
    try {
      const raw = await GetTypes({ source_key: sourceKey })
      let arr: any[] = []
      if (Array.isArray(raw)) arr = raw
      else if (raw && typeof raw === 'object') {
        if (Array.isArray((raw as any).list)) arr = (raw as any).list
        else if ((raw as any).type_id && (raw as any).name) arr = [raw]
      }
      types.value = arr.map((t: any) => ({ type_id: t?.type_id ?? '', name: t?.name ?? '' }))
    } catch (e: any) {
      errorStore.fromError('加载分类失败', e, 'videoStore.loadTypes')
      types.value = []
    }
  }

  async function loadYearsAndAreas(sourceKey: string): Promise<void> {
    try {
      const resp = (await GetYearsAndAreas(sourceKey)) as any
      years.value = Array.isArray(resp?.years) ? resp.years : []
      areas.value = Array.isArray(resp?.areas) ? resp.areas : []
    } catch (e: any) {
      errorStore.fromError('加载年份/地区选项失败', e, 'videoStore.loadYearsAndAreas')
      years.value = []
      areas.value = []
    }
  }

  async function loadRecommend(sourceKey: string, excludeIds: string[], limit = 12): Promise<Video[]> {
    try {
      const list = (await GetRecommend({
        source_key: sourceKey,
        limit,
        exclude_ids: excludeIds,
      })) as any
      return Array.isArray(list) ? list : []
    } catch (e: any) {
      errorStore.fromError('加载推荐失败', e, 'videoStore.loadRecommend')
      return []
    }
  }

  return {
    videos,
    currentVideo,
    episodes,
    types,
    years,
    areas,
    total,
    page,
    loading,
    loadVideos,
    loadDetail,
    refreshDetail,
    search,
    loadTypes,
    loadYearsAndAreas,
    loadRecommend,
    deletedVodId,
    deletedSourceKey,
    notifyDeletion,
    clearDeletionNotify,
    refreshTrigger,
    notifyRefresh,
  }
})
